subprojects/core/src/main/java/org/gradle/api/internal/tasks/DefaultTaskInputs.java

```diff
@@ -84,10 +85,11 @@ public void visitRegisteredProperties(PropertyVisitor visitor) {
             visitor.visitInputFileProperty(
                 registration.getPropertyName(),
                 registration.isOptional(),
-                registration.isSkipWhenEmpty(),
+                registration.isSkipWhenEmpty()
```

Comment:
(I don't know why I can't see this comment when looking at the changes in the PR...)

The difference is that we now allow runtime-registered `skipWhenEmpty(true)` input properties to be queried via `InputChanges`, and previously we didn't.

To actually observe a difference you'd need the following:

1) an incremental task, i.e. a task that has a `@TaskAction` method taking `IncrementalInputs` as a parameter; it is not possible to register an incremental task using the runtime API,
2) the task should register its own `skipWhenEmpty()` inputs via the runtime API; this is allowed, but it's a strange behavior (more on this later),
3) it would have to query changes to these inputs through `InputChanges`.

```groovy
abstract class MySourceTask extends DefaultTask {
    @Internal abstract ConfigurableFileCollection getSources()

    MySourceTask() {
        inputs.files(sources).skipWhenEmpty()
    }

    @TaskAction
    void execute(InputChanges changes) {
        changes.getFileChanges(sources).each { ... }
    }
}
```

So after the changes in the PR you get the actual changes from `getFileChanges(sources)`, and previously you'd have gotten `null`.

This is still a behavior change, but it's not a breaking one. The pattern you'd need to get here is definitely not recommended, and indeed we might want to disallow calling `TaskInputs` from the task constructor.

I see a number of ways to deal with this:

1) deprecate `TaskInputFilePropertyBuilder.skipWhenEmpty()` because the use case for it is not terribly strong in the first place (we use it in `AbstractCopyTask` but only because we are lacking some features in `@Nested` handling, and are forced to do some weird stuff; we could fix this, or switch to doing the weird stuff via an internal-only API if we have to,
2) allow people to declare runtime properties as `incremental()` as well, to bring the functionality between the runtime and the annotation API into alignment,
3) or ignore this completely.

I'm leaning towards 3), because a) it's unlikely that anyone will ever use this feature, and b) if they do and they actually need 2), it is easy enough to add it later.

Meanwhile we should remove the need for `AbstractCopyTask` to call the runtime API, and perhaps deprecate the use of the runtime API from task constructors entirely.